strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,957 @@
+        covmat = self.hbar * np.diag(np.array([eps ** 2, 1.0 / eps ** 2])) / 2
+        return self.measure_dyne(covmat, [mode], shots=shots)
+
+    def heterodyne(self, mode, shots=1):
+        r"""Performs a heterodyne measurement on a mode, simulated by a generaldyne
+        onto a coherent state.
+
+        Args:
+            mode (int): mode to be measured
+            shots (int): how many measurements are performed
+
+        Returns:
+            array: heterodyne outcome
+        """
+        covmat = self.hbar * np.eye(2) / 2
+        return self.measure_dyne(covmat, [mode], shots=shots)
+
+    def post_select_generaldyne(self, covmat, modes, vals):
+        r"""Simulates general-dyne measurement on a set of modes with a specified measurement
+        outcome.
+
+        Args:
+            covmat (array): covariance matrix of the generaldyne measurement
+            modes (list): modes to be measured
+            vals (array): measurement outcome to postselect
+
+        Raises:
+            ValueError: if the dimension of covmat does not match number of quadratures
+                        associated with modes
+            ValueError: if any of the modes are not in the list of active modes
+        """
+        if covmat.shape != (2 * len(modes), 2 * len(modes)):
+            raise ValueError("Covariance matrix size does not match indices provided")
+
+        for i in modes:
+            if self.active[i] is None:
+                raise ValueError("Cannot apply measurement, mode does not exist")
+
+        expind = np.concatenate((2 * np.array(modes), 2 * np.array(modes) + 1))
+        mp = self.get_covmat()
+        A, B, C = ops.chop_in_blocks_multi(mp, expind)
+        V = A - B @ np.linalg.inv(C + covmat) @ B.transpose(0, 2, 1)
+        self.covs = ops.reassemble_multi(V, expind)
+
+        r = self.get_mean()
+        (va, vc) = ops.chop_in_blocks_vector_multi(r, expind)
+        va = va + np.einsum("...ij,...j", B @ np.linalg.inv(C + covmat), (vals - vc))
+        self.means = ops.reassemble_vector_multi(va, expind)
+
+        # Reweight each peak based on how likely a given peak was to have
+        # contributed to the observed outcome
+        reweights_exp_arg = np.einsum(
+            "...j,...jk,...k", (vals - vc), np.linalg.inv(C + covmat), (vals - vc)
+        )
+        reweights = np.exp(-0.5 * reweights_exp_arg) / (
+            np.sqrt(np.linalg.det(2 * np.pi * (C + covmat)))
+        )
+        self.weights *= reweights
+        self.weights /= np.sum(self.weights)
+
+        self.means = self.means[abs(self.weights) > 0]
+        self.covs = self.covs[abs(self.weights) > 0]
+        self.weights = self.weights[abs(self.weights) > 0]
+
+    def post_select_homodyne(self, mode, val, eps=0.0002, phi=0):
+        r"""Simulates a homodyne measurement on a mode, postselecting on an outcome.
+
+        Args:
+            mode (int): mode to be measured
+            val (array): measurement value to post-select
```

Comment:
Seems to have both ` post-select` and ` postselect` within the file, would suggest going with only one

Reply:
Will do